lib/src/auth/gateway/auth_gateway.dart

```diff
@@ -56,9 +58,27 @@ class AuthGateway {
     );
 
     if (response.statusCode != 200) {
-      throw Exception('${response.statusCode}: ${response.reasonPhrase}');
+      if (response.body == null) {
+        throw AuthException(
+          response.statusCode.toString(),
+          response.reasonPhrase,
+        );
+      }
+      final responseBody = json.decode(response.body) as Map<String, dynamic>;
+      final code = _parseCodeFromJson(responseBody);
+      throw AuthException.fromCode(code);
     }
 
     return json.decode(response.body) as Map<String, dynamic>;
   }
+
+  String _parseCodeFromJson(Map<String, dynamic> json) {
```

Comment:
Is it possible for the given `json` to be `null`, and should we handle this case?